### PR TITLE
add docker support

### PR DESCRIPTION
**Pull request by ldoublewood**
1. Add Dockerfile file, mostly borrow from ipfs project
2. Add .dockerignore file
3. Modify Makefile to build extern submodule 'filecoin-ffi' from source instead of download from github release

This PR can close issue:  #594

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,7 +22,7 @@ FFI_DEPS:=$(addprefix $(FFI_PATH),$(FFI_DEPS))
 $(FFI_DEPS): build/.filecoin-install ;
 
 build/.filecoin-install: $(FFI_PATH)
-	$(MAKE) -C $(FFI_PATH) $(FFI_DEPS:$(FFI_PATH)%=%)
+	FFI_BUILD_FROM_SOURCE=1 $(MAKE) -C $(FFI_PATH) $(FFI_DEPS:$(FFI_PATH)%=%)
```
Comment (whyrusleeping) [inline Makefile]: don't commit this line

Comment (ldoublewood) [reply Makefile]: restore Makefile done

**Comment by whyrusleeping** [comment]
Gonna ask @Kubuxu to do the review here, i'm not good at dockerfiles

**Comment by eshon** [comment]
In case it helps @travisperson shared these Docker scripts. They've been edited though and not tested after edit so just pasting here for comparison. 

https://gist.github.com/travisperson/1859967cdb7abb5f020289b578528edf

This one doesn't expose port 5678

**Comment by whyrusleeping** [comment]
Thanks @eshon! Thats good context to have. I'm tagging @travisperson for review now

**Comment by eshon** [comment]
Note that if people have an earlier install they might have docker cache problems so some instructions running clean versions would be great as comments @ldoublewood

**Comment by ldoublewood** [comment]
@eshon Would you please elaborate what kind of "docker cache problem" will happen?

**Comment by eshon** [comment]
I think if you don't clear your cache it may not pull latest from master but from a saved copy in your cache so may want to run with `--no-cache` option or whatever else there is, someone on Slack had that problem.

**Comment by whyrusleeping** [comment]
@eshon should we add anything to the docs about this? I think we can merge this now in any case

**Comment by eshon** [comment]
For now, you guys could create a `tools/dockers/docker-examples` directory in the lotus repo for community-contributed dockers and docker compose scripts.

@travisperson could use `tools/dockers/dockers-infra` but let's see what he thinks.

Each community contribution should have a separate descriptive folder of its purpose. For example: 

- `tools/dockers/docker-examples/basic-miner-busybox/` (for ldoublewood's here)
- `tools/dockers/docker-examples/dev-api-arch` (for [mateo's](https://github.com/mateodelnorte/filecoin-lotus-docker) docker compose)
- `tools/dockers/docker-examples/new-network-busybox` (for a potential local devnet Docker)

`tools/dockers/README.md` says "`docker-examples` are community-contributed. Have a new type of Docker to share? Make a PR with your scripts or a README.md linking to its home repo and purpose. See a bug? File an Issue so we can keep these up-to-date."

---------

**TLDR;**

There are different Docker and Docker Compose scripts for different use cases. 

Use Cases:

1. Spinning up a miner
    a. basic storage miner (ldoublewood's here is a good starter for this?)
    b. distributed storage miner (docker compose >> api for seal workers, etc.)
2. Spinning up API endpoints for building apps 
    a. local node for a developer
    b. hosted semi-public endpoint for multiple developers 
3. Spinning up a local network or new network
    a. basic local network (e.g. https://lotu.sh/en+setup-local-dev-net)
    b. new devnet (faucet, genesis, bootstrap, etc.)

Nice-to-have:
- optional caching of proof params

The dockers in go-filecoin's `.docker` directory and ones that @travisperson creates are traditionally for use case 3b. They are regularly used and tested by Infra.

However it'd be nice to collect Dockers for all use cases somewhere for the community that's easy to find.

_Option 1) Community dockers in a new Docker directory in the lotus repo_ - Easy for community members to find.

_Option 2) filecoin-shipyard/lotus-dockers_ or _filecoin-project/dockers_ - Less visibility, far from node implementation but in tune with go-filecoin's tradition that repo Dockers are for network infra so just references and actively maintained.

TODO:
Ask Jim to add Docker links to the docs wherever Dockers end up

**Comment by jsign** [comment]
A quick run of these Dockerfile seemed to have non-growing volumes. Taking a look inside the container `.lotus` was living on `/root`. 
Coming back to the Dockerfile:
- `USER lotus` seems to be missing just before `CMD`
- Shouldn't `PARAMCACHE_PATH` be `FIL_PROOFS_PARAMETER_CACHE`?

**Comment by jsign** [comment]
Running with `USER lotus` has some small problem with `/var/tmp/...` permissions since `lotus daemon` fails while downloading and loading the params. The folder is owned by `root` and `users` group doesn't have enough permissions! so an extra tweak seems to be needed there.

Maybe some anyone can further check if these facts are correct :)

**Comment by whyrusleeping** [comment]
Gonna merge this now, lots of people have been asking for it. Its probably not perfect yet, but we *do* accept PRs :)
